flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java

```diff
@@ -242,21 +239,22 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
 				throw new IOException(t);
 			}
 			finally {
-				if (buffer != null) {
-					buffer.recycle();
-				}
+				buffer.recycle();
 			}
 		}
 
-		@Override
-		void readFrom(ByteBuf buffer) {
-			receiverId = InputChannelID.fromByteBuf(buffer);
-			sequenceNumber = buffer.readInt();
-			isBuffer = buffer.readBoolean();
-			size = buffer.readInt();
-
-			retainedSlice = buffer.readSlice(size);
-			retainedSlice.retain();
+		static BufferResponse readFrom(ByteBuf buffer) {
+			BufferResponse result = new BufferResponse();
```

Comment:
this is somehow strange construct, with this default empty constructor and manually modifying fields afterwards. Why not:
```
InputChannelID receiverId = InputChannelID.fromByteBuf(buffer);
int sequenceNumber = buffer.readInt();
boolean isBuffer = buffer.readBoolean();
int size = buffer.readInt();

return new BufferResponse(
  receivedId,
  sequenceNumber,
  isBuffer,
  size,
  buffer.readSlice(size).retain());
```

would be more natural, would be easier for the finding usages of the fields and that way you could make all of the fields `final`, which is nice feature on its own.

Reply:
I agree we could change that, too - and partly I did in another commit that didn't go into a PR yet. Let's do this for all classes then.